Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -384,4 +356,269 @@ static void _GetGlyphRunDetails(_CTTypesetter* ts, CFRange range, CGRect frameSi
+static ComPtr<IDWriteFontFamily> __GetDWriteFontFamily(NSString* familyName) {
+    ComPtr<IDWriteFontFamily> fontFamily;
+
+    ComPtr<IDWriteFactory> factory = __GetDWriteFactoryInstance();
+
+    ComPtr<IDWriteFontCollection> systemFontCollection;
+    THROW_IF_FAILED(factory->GetSystemFontCollection(&systemFontCollection));
+
+    NSUInteger familyNameLength = familyName.length;
+    std::vector<UniChar> unicharFamilyName(familyNameLength + 1);
+    [familyName getCharacters:unicharFamilyName.data() range:NSMakeRange(0, familyNameLength)];
+
+    UINT32 fontFamilyIndex;
+    BOOL fontFamilyExists;
+
+    THROW_IF_FAILED(systemFontCollection->FindFamilyName(reinterpret_cast<const wchar_t*>(unicharFamilyName.data()),
+                                                         &fontFamilyIndex,
+                                                         &fontFamilyExists));
+
+    if (!fontFamilyExists) {
+        TraceError(TAG, L"Unable to find font family \"%ws\"", unicharFamilyName.data());
+        return fontFamily;
+    }
+
+    THROW_IF_FAILED(systemFontCollection->GetFontFamily(fontFamilyIndex, &fontFamily));
+    return fontFamily;
+}
+
+// Private helper that parses a font name, and returns appropriate weight, stretch, and style values for DWrite accordingly
+static void __InitDWriteFontPropertiesFromName(CFStringRef fontName,
+                                               DWRITE_FONT_WEIGHT* weight,
+                                               DWRITE_FONT_STRETCH* stretch,
+                                               DWRITE_FONT_STYLE* style) {
+    // Set some defaults for when weight/stretch/style are not mentioned in the name
+    *weight = DWRITE_FONT_WEIGHT_NORMAL;
+    *stretch = DWRITE_FONT_STRETCH_NORMAL;
+    *style = DWRITE_FONT_STYLE_NORMAL;
+
+    NSString* familyName = _DWriteGetFamilyNameForFontName(fontName);
+
+    // Relationship of family name -> font name not always consistent
+    // Usually, properties are added to the end (eg: Arial -> Arial Narrow Bold)
+    // However, this is not always the case (eg: Eras ITC -> Eras Bold ITC)
+    // In addition, some fonts with properties are occasionally placed into their own family (eg: Segoe WP SemiLight -> Segoe WP SemiLight)
+    // Try to be more prudent about these edge cases, by looking only at the difference between the font name and family name
+    NSArray<NSString*>* fontNameTokens = [(__bridge NSString*)fontName componentsSeparatedByString:@" "];
+    NSMutableSet<NSString*>* propertyTokens = [NSMutableSet<NSString*> setWithArray:fontNameTokens];
+
+    if (familyName) {
+        NSArray<NSString*>* familyNameTokens = [familyName componentsSeparatedByString:@" "];
+        [propertyTokens minusSet:[NSSet setWithArray:familyNameTokens]];
+    }
+
+    for (NSString* propertyToken in propertyTokens) {
+        NSString* upperPropertyToken = [propertyToken uppercaseString];
+
+        if ([upperPropertyToken isEqual:@"THIN"]) {
```

Comment:
probably.  I imagine lookup would be faster than all of the string comparisons it would take to be italic

---
In reply to: [78228119](https://github.com/Microsoft/WinObjC/pull/987#discussion_r78228119) [](ancestors = 78228119)